### PR TITLE
[RWRoute] Signal router for Versal architecture

### PR DESCRIPTION
**Pull request by WenhaoLin-AMD**
Make the signal router in RWRoute support Versal architecture.

The main changes needed to extend beyond UltraScale/UltraScale+ support:
- New Versal tile types `CLE_BC_CORE` and `INTF_[LR]OCF_[TB][LR]_TILE`
- Changed Versal node intent codes in
  - `INT` tiles
    - `NODE_SINGLE` (UltraScale/UltraScale+) -> `NODE_[VH]SINGLE` (Versal)
    - `NODE_DOUBLE` -> `NODE_[VH]DOUBLE`
    - `NODE_HLONG` -> `NODE_HLONG[6, 10]`
    - `NODE_VLONG` -> `NODE_VLONG[7, 12]`
    - (new types) `NODE_INODE`, `NODE_IMUX`, `NODE_SDQNODE`
  - `CLE_BC_CORE` tiles
    - `NODE_CLE_CTRL`
    - `NODE_CLE_[BC]NODE`
  - `INTF_[LR]OCF_[TB][LR]_TILE` tiles
    - `NODE_INTF_CTRL`
    - `NODE_INTF_[BC]NODE`

Not currently supported in Versal:
- SLR crossings
- `--lutPinSwapping`
- `--lutRoutethru`

### CODE REVIEW

File: src/com/xilinx/rapidwright/rwroute/RouteNode.java [hunk 1]
```diff
@@ -169,13 +174,43 @@ private void setBaseCost() {
...
+                        baseCost = 0.15f * minLength;
+                        break;
                     case NODE_HLONG:
                         assert (length != 0 || getAllDownhillNodes().isEmpty());
                         baseCost = 0.15f * length;// HLONGs have length 6 and 7
                         break;
+                    case NODE_VLONG7:  // Versal only
+                    case NODE_VLONG12: // Versal only
                     case NODE_VLONG:
                         baseCost = 0.7f;
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/rwroute/RouteNode.java:189]: Should VLONG7 and VLONG12 have the same base cost?

**Review by eddieh-xlnx — COMMENTED** [review]
Another area that needs some work is that we want to update the following output to count Versal node types:
```
==============================================================================
==                                Statistics                                ==
==============================================================================
Node Usage Per Type
 Node Type                 Usage        Length
 NODE_SINGLE                   0             0
 NODE_DOUBLE                   0             0
 NODE_VQUAD                   22            37
 NODE_HQUAD                   27            54
 NODE_VLONG                    0             0
 NODE_HLONG                    0             0
 NODE_LOCAL                    0             0
 NODE_PINBOUNCE              257             0
 NODE_PINFEED                  0             0
 NODE_LAGUNA_DATA              0             0
```

File: src/com/xilinx/rapidwright/rwroute/RouterHelper.java [hunk 2]
```diff
@@ -143,7 +144,7 @@ public static boolean isExternalConnectionToCout(SitePinInst source, SitePinInst
      * @return A node that connects to an INT tile from an output pin.
      */
     public static Node projectOutputPinToINTNode(SitePinInst output) {
-        int watchdog = 5;
+        int watchdog = 20;
```
[the remark this answers is not in the document]
Comment (WenhaoLin-AMD) [reply src/com/xilinx/rapidwright/rwroute/RouterHelper.java:148]: It's a very early fix for corescore_500. I reverted this change and got:

> ...
> 
> INFO: Route 3 direct connections 
> ERROR: Failed to find a path between two nodes: XPIO_NIBBLE_SC_1_X9Y0/XPIO_IOBPAIR_5_RXOUT_M_PIN, CMT_MMCM_X11Y0/CMT_MMCM_TOP_0_CLKIN1_PIN
> ERROR: Failed to route direct connection Con 327492, boundbox = [( 0, 0 ), -> ( 0, 0 )], net = i_clk, net fanout =   1, source = RXOUT_M, sink = CLKIN1, delay =    0 , criticality = 0.000 
> ERROR: Failed to find a path between two nodes: CMT_MMCM_X11Y0/CMT_MMCM_TOP_0_LOCKED_PIN, CLE_E_CORE_X26Y144/CLE_SLICEL_TOP_0_AX_PIN
> ERROR: Failed to route direct connection Con 458939, boundbox = [( 0, 0 ), -> ( 0, 0 )], net = clock_gen/locked, net fanout =   1, source = LOCKED, sink = AX, delay =    0 , criticality = 0.000 
> ERROR: Failed to find a path between two nodes: CMT_MMCM_X11Y0/CMT_MMCM_TOP_0_CLKOUT0_PIN, CLK_REBUF_BUFGS_HSR_CORE_X8Y0/CLK_BUFGCE_59_I_PIN
> ERROR: Failed to route direct connection Con 597540, boundbox = [( 0, 0 ), -> ( 0, 0 )], net = clock_gen/clk, net fanout =   1, source = CLKOUT0, sink = I, delay =    0 , criticality = 0.000 
> 
> INFO: No PIP overlaps
> 
> ...

If the value of `watchdog` is too small, some of the connections that should be classified as indirect will be classified as direct.

**Comment by WenhaoLin-AMD** [comment]
> Another area that needs some work is that we want to update the following output to count Versal node types:
> 
> ```
> ==============================================================================
> ==                                Statistics                                ==
> ==============================================================================
> Node Usage Per Type
>  Node Type                 Usage        Length
>  NODE_SINGLE                   0             0
>  NODE_DOUBLE                   0             0
>  NODE_VQUAD                   22            37
>  NODE_HQUAD                   27            54
>  NODE_VLONG                    0             0
>  NODE_HLONG                    0             0
>  NODE_LOCAL                    0             0
>  NODE_PINBOUNCE              257             0
>  NODE_PINFEED                  0             0
>  NODE_LAGUNA_DATA              0             0
> ```

Has updated the above part in https://github.com/Xilinx/RapidWright/pull/1077/commits/cc2e98af6af653e1bfb23bb417785919ade98056

File: src/com/xilinx/rapidwright/rwroute/RouteNode.java (hunk 1 above)
Comment (WenhaoLin-AMD) [reply src/com/xilinx/rapidwright/rwroute/RouteNode.java]: Seems that assign the base cost according to the length of a node is better. Has updated this part in https://github.com/Xilinx/RapidWright/pull/1077/commits/06658011533dd7ff8f2ce07d97c52541d2322932

File: src/com/xilinx/rapidwright/rwroute/RouterHelper.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (eddieh-xlnx) [reply src/com/xilinx/rapidwright/rwroute/RouterHelper.java:148]: Got it, thanks. It's a deep path to get from IOB to the fabric!

File: src/com/xilinx/rapidwright/rwroute/PartialRouter.java [hunk 3]
```diff
@@ -342,6 +346,8 @@ protected void addNetConnectionToRoutingTargets(Net net) {
             }
 
             if (net.hasPIPs()) {
+                final boolean isVersal = (design.getSeries() == Series.Versal);
```
Comment (clavin-xlnx) [inline src/com/xilinx/rapidwright/rwroute/PartialRouter.java:349]:
```suggestion
                final boolean isVersal = design.getSeries() == Series.Versal;
```

**Review by clavin-xlnx — APPROVED** [review]

File: src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java [hunk 4]
```diff
@@ -517,7 +534,17 @@ public int numNodes() {
     }
 
     protected RouteNode create(Node node, RouteNodeType type) {
-        return new RouteNode(this, node, type);
+        RouteNode rnode = new RouteNode(this, node, type);
+        // PINFEED_I should have zero length, except for on US/US+ where the PINFEED_I is a PINBOUNCE node.
```
Comment (eddieh-xlnx) [inline src/com/xilinx/rapidwright/rwroute/RouteNodeGraph.java:538]:
```suggestion
        // PINFEED_I should have zero length, except for on US/US+ where the PINFEED_I can be a PINBOUNCE node.
```

File: test/src/com/xilinx/rapidwright/rwroute/TestRWRoute.java [hunk 5]
```diff
@@ -293,6 +293,63 @@ public void testNonTimingDrivenPartialRouting() {
...
+                if (net.getName().equals("processor/address_loop[6].output_data.pc_vector_mux_lut/O6")) {
+                    // For one hand chosen net, block all its downhill PIPs such that
+                    // its output pin
+                    net.unroute();
+                    Node sourcePinNode = net.getSource().getConnectedNode();        // CLE_W_CORE_X50Y6/CLE_SLICEL_TOP_0_D_O_PIN
+                    Node sourceNode = sourcePinNode.getAllDownhillNodes().get(0);   // CLE_W_CORE_X50Y6/CLE_SLICEL_TOP_0_D_O
+                    design.setTrackingChanges(false);
+                    for (PIP pip : sourceNode.getAllDownhillPIPs()) {
+                        Z_NET.addPIP(pip);
+                    }
```
Comment (eddieh-xlnx) [inline test/src/com/xilinx/rapidwright/rwroute/TestRWRoute.java:323]:
```suggestion
                    // For one hand chosen net, unroute it entirely and block
                    // all of its primary output's downhill PIPs such that 
                    // the alternate output must be used
                    net.unroute();
                    Node sourcePinNode = net.getSource().getConnectedNode();        // CLE_W_CORE_X50Y6/CLE_SLICEL_TOP_0_D_O_PIN
                    Node sourceNode = sourcePinNode.getAllDownhillNodes().get(0);   // CLE_W_CORE_X50Y6/CLE_SLICEL_TOP_0_D_O
                    design.setTrackingChanges(false);
                    for (PIP pip : sourceNode.getAllDownhillPIPs()) {
                        Z_NET.addPIP(pip);
                    }
```

**Pull request merged**